### DISCUSSION TITLE
Summing & Grouping in BB Database

### DISCUSSION THREAD

**Discussion by philnorton** [Budibase/budibase#3933] [Help] [answered]
Hey, just came across Budibase today, and it seems really interesting! Been playing with the layouts etc, and thought I'd create a very simple expenses tracker. So, I've got 2 tables: Expenses and Categories. An Expense can have 1 Category, and 1 Category can have many Expenses. 

So far, so simple... Now, I may be guilty of not fully RTFM, but I want to be able to calculate things like... Total Expenses by Category. The amount for each item is stored as a Number in Expenses, and I've tried to create a View in the Expenses table to group by Category, but that doesn't seem possible? Am I thinking about this all wrong?

Next thing I've tried is to create a formula column in the Categories, with Total Spend, but I'm not entirely sure how to get that to sum all amounts in that category? I'm _assuming_ I'd need to use JS or Handlebars?

So I guess my question is, what's the best practise for this type of calculation?

**Comment by mjashanks** [comment]
Hey Phil. Welcome!

At the moment, it's not possible as long as you are storing your Categories in a table. The way we would like this to work is by using Views, which are supposed to create subsets of table data, including grouped calculations. A limitation of view is that you cannot group by relationships. I.e. we would like to be able to create a view on the `Expenses` table, grouped by `Category`.

Here is a list of View limitations/upgrades that we would like to complete soon: https://github.com/Budibase/budibase/discussions/2999

What you could do is make your expenses `Category` column a `Text` or `Options` type - i.e. not a relationship. You could then setup a view that is grouped by category. Views docs: https://docs.budibase.com/building-apps/data/views

You could still have a category table, but "soft link" it to Expenses, when needed - i.e. by the Category Name and the Category Text column.

Regarding your formula question - again you should use views for this. A formula field only has access to data within the same row.

**Reply by CovantaDataManagement** [reply]
I am trying to get a total of timerecord detail records for an individual worker to save a "Total Hours sum" for a defined time period.  Is there a means to do this?

**Reply by mjashanks** [reply, marked as answer]
@CovantaDataManagement this is possible, but you need to write a bit of Javascript.

You will need the following component structure:
```
> Data Provider : Worker
  > Repeater
     .....worker details
        > Data Provider "Time Card Provider"  : Relationships > Worker Repeater.TimeCardDetail
              > Paragraph : With Javascript binding below that counts the hours
```

Javascript that will sum up the hours on your "Time Card Provider"

```javascript
let sum = 0;
for(let timecard of $("Timecard Provider.Rows")) {
	sum = sum + timecard.Hours;
}
return sum;
```
Here is a video of what mine looks like. I am using `Contacts`, who have many `Invoices`:

https://user-images.githubusercontent.com/3524181/148962040-b9af8df8-6c5b-4127-9318-7e427ebcd92b.mp4

**Reply by gerhardmpl** [reply]
@mjashanks In your example: would it be possible to also calculate the total sum of hours for all workers? From what I understand, the javascript will only calculate the hours for one worker per repeater loop. How can you pass the total hours to a field in a screen?

**Reply by mjashanks** [reply]
@gerhardmpl  - if you are using Budibase DB, you might be better using a view for this: https://docs.budibase.com/docs/views

**Reply by krugmer** [reply]
Hi, 
thanks a lot for your post, it helped me a lot in understanding the potential of JavaScript in paragraphs for my little project.
I am facing an issue and I hope someone may be able to help. 
I have teachers reporting their contact hours for the modules they teach. Each raw in the "hours" table is an entry (ie hours taught I a module divided by theoretical, practical and revision).
I would like to obtain the sum of hours taught by each teacher (ie sum to theoretical hours for teacher x, y, z..)
I structured my providers and repeater similarly to what was described.
I have:
- a table that contains the teachers
- a table that contains the hours entry for each teacher

```
> Data Provider : teacher
  > Repeater
     .....teacher details
        > Data Provider "hours thaught"  : Relationships > teacher Repeater.hours details
              > Paragraph : With Javascript binding below that counts the hours
```

 I did manage to obtain a list of teachers with the entries they made and the following code work:

`return $("Hours Provider.Rows")`
successfully returns the the objects for each teacher (ie [object Object],[object Object],[object Object])

`return $("Hours Provider.Rows.1.Hours theoretical") `
successfully returns the number of hours of entry "1" (first object in the array) for each teacher 

However when I try to sum up the hours... 

```
let sum = 0; 
for(let entry of $("Hours Provider.Rows")) { 
sum = sum + entry.Hours theoretical; 
} 
return sum;
```
here is where I get stuck, there is an error that I am not able to fix, any suggestions?
Many thanks and apologies for the newbie question!

**Reply by krugmer** [reply]
Ok, I finally managed to solve the issue and it turned out to be a problem with the variable name that had a space(!)
I solved it by using the backticks (`) and square brackets to encapsulate the variable name. (see below)
Thanks to Bard for the tip 

```
const hourslogs = $("Hours Provider.Rows");
let totalhourstaught = 0;
let i=0;
for (let i in hourslogs) {
  totalhourstaught = totalhourstaught+hourslogs[i][`Hours theoretical`];
}
return "Hours taught: " + totalhourstaught;
```

**Comment by andyburgessmd** [comment]
Did something happen to views? I swear you used to be able to do things like sum and 'group by' but don't see those options anymore.

**Reply by ConorWebb96** [reply]
The functionality was removed, once SQS is released and additional amends are made to it, this functionality should return.

**Reply by tneugebauer** [reply]
@ConorWebb96 its a pitty that this particular functionality is removed from the views, as one isn´t able to use it anymore. What is the (best) workaround to sum a row or column of data?

**Reply by Shadowfita** [reply]
I am chasing this too!

**Reply by Shadowfita** [reply]
@tneugebauer my workaround has been to call the internal API with a REST API request and use the javascript transformer to group how I need it to be grouped.

**Reply by Shadowfita** [reply]
@tneugebauer 

```

function countOccurrencesBy(groupBy, jsonResponse) {
    const countsByTopic = {};

    jsonResponse.data.forEach(item => {
        const topic = item[groupBy];

        if (!countsByTopic[topic]) {
            countsByTopic[topic] = 0;
        }

        countsByTopic[topic]++;
    });

    const output = Object.keys(countsByTopic).map(topic => {
        return { "Label": topic, "Count": countsByTopic[topic] };
    });

    return output;
}


return countOccurrencesBy("Topic", data);

```

**Reply by unknownkwita** [reply]
What is "SQS"?

I would prefer not to implement the workaround if the fix is on the horizon.

**Reply by ConorWebb96** [reply]
Hey @unknownkwita,

SQS will allow you to write queries similar to MySQL, postgres etc for the internal Budibase tables. This should allow you to write a query that would group your data.

We don't have a current ETA on SQS as the engineering team have been pulled off to work on important bugs.

**Reply by Beethoven9** [reply]
+1...I mean, that functionality was asewome and, IMHO, core for a low-code app.
